### PR TITLE
refact(draw_vector): standardized draw gradient API

### PR DESCRIPTION
**Pull request by FASTSHIFT**
### Description of the feature or fix

Reconstructs the vector draw gradient interface, aligned with the SVG standard. No coordinate conversion is required.

TODO:

- [x] Fix software rendering errors.
- [x] Adapts the new interface to VG-Lite porting.
- [x] Implement `vg_lite_draw_linear_grad` in the VG-Lite simulator.

cc @onecoolx 

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) version [v3.4.12](https://github.com/szepeviktor/astyle/releases/tag/v3.4.12) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

**Review by onecoolx — APPROVED** [review]

File: tests/ref_imgs_vg_lite/draw/render/argb8888/demo_render_conical_gradient_opa_128.png
Comment (kisvegabor) [inline tests/ref_imgs_vg_lite/draw/render/argb8888/demo_render_conical_gradient_opa_128.png:1]: FYI, these ref images are broken.

Comment (FASTSHIFT) [reply tests/ref_imgs_vg_lite/draw/render/argb8888/demo_render_conical_gradient_opa_128.png:1]: This is a known issue. Due to the tricky coordinate conversion issues involved, the VG-Lite backend does not yet support these parameters in `lv_grad_dsc_t`:
```c
typedef struct {
    XXX
#if LV_USE_DRAW_SW_COMPLEX_GRADIENTS
    union {
        /*Linear gradient parameters*/
        struct {
            lv_point_t  start;                          /**< Linear gradient vector start point */
            lv_point_t  end;                            /**< Linear gradient vector end point */
        } linear;
        /*Radial gradient parameters*/
        struct {
            lv_point_t  focal;                          /**< Center of the focal (starting) circle in local coordinates */
            /* (can be the same as the ending circle to create concentric circles) */
            lv_point_t  focal_extent;                   /**< Point on the circle (can be the same as the center) */
            lv_point_t  end;                            /**< Center of the ending circle in local coordinates */
            lv_point_t  end_extent;                     /**< Point on the circle determining the radius of the gradient */
        } radial;
        /*Conical gradient parameters*/
        struct {
            lv_point_t  center;                         /**< Conical gradient center point */
            int16_t     start_angle;                    /**< Start angle 0..3600 */
            int16_t     end_angle;                      /**< End angle 0..3600 */
        } conical;
    } params;
    void * state;
#endif
} lv_grad_dsc_t;
```

So don't implement this drawing for now:
```c
bool lv_vg_lite_draw_grad_helper(
    struct _lv_draw_vg_lite_unit_t * u,
    vg_lite_buffer_t * buffer,
    vg_lite_path_t * path,
    const lv_area_t * area,
    const lv_grad_dsc_t * grad_dsc,
    const vg_lite_matrix_t * matrix,
    vg_lite_fill_t fill,
    vg_lite_blend_t blend)
{
    LV_ASSERT_NULL(u);
    LV_VG_LITE_ASSERT_DEST_BUFFER(buffer);
    LV_VG_LITE_ASSERT_PATH(path);
    LV_ASSERT_NULL(area);
    LV_ASSERT_NULL(grad_dsc);
    LV_VG_LITE_ASSERT_MATRIX(matrix);

    lv_vector_gradient_t grad;
    lv_memzero(&grad, sizeof(grad));

    grad.style = LV_VECTOR_GRADIENT_STYLE_LINEAR;
    grad.spread = LV_VECTOR_GRADIENT_SPREAD_PAD;
    grad.stops_count = grad_dsc->stops_count;
    lv_memcpy(grad.stops, grad_dsc->stops, sizeof(lv_gradient_stop_t) * grad_dsc->stops_count);

    switch(grad_dsc->dir) {
        case LV_GRAD_DIR_VER:
            grad.x1 = area->x1;
            grad.y1 = area->y1;
            grad.x2 = area->x1;
            grad.y2 = area->y2 + 1;
            break;

        case LV_GRAD_DIR_HOR:
            grad.x1 = area->x1;
            grad.y1 = area->y1;
            grad.x2 = area->x2 + 1;
            grad.y2 = area->y1;
            break;

        default:
            LV_LOG_WARN("Unsupported gradient direction: %d", grad_dsc->dir);
            return false;
    }

    vg_lite_matrix_t grad_matrix;
    vg_lite_identity(&grad_matrix);

    return lv_vg_lite_draw_grad(u, buffer, path, &grad, &grad_matrix, matrix, fill, blend);
}
```

Comment (FASTSHIFT) [reply tests/ref_imgs_vg_lite/draw/render/argb8888/demo_render_conical_gradient_opa_128.png:1]: It is not possible to implement all the functions defined by `LV_USE_DRAW_SW_COMPLEX_GRADIENTS`. I have implemented the supported parts and the comparison images have been updated.

File: demos/vector_graphic/lv_demo_vector_graphic.c [hunk 1]
```diff
@@ -92,20 +90,18 @@ static void draw_radial_gradient(lv_vector_dsc_t * ctx, lv_vector_path_t * path)
...
+    lv_gradient_stop_t stops[2];
+    lv_memzero(stops, sizeof(stops));
+    stops[0].color = lv_color_hex(0xff0000);
+    stops[0].opa = LV_OPA_COVER;
+    stops[0].frac = 0;
+    stops[1].color = lv_color_hex(0x0000ff);
+    stops[1].opa = LV_OPA_COVER;
+    stops[1].frac = 255;
+
+    lv_vector_dsc_set_fill_radial_gradient(ctx, 450, 100, 20);
```
Comment (liamHowatt) [inline demos/vector_graphic/lv_demo_vector_graphic.c:102]: The radial gradient `cx` and `cy` are absolute coordinates now?

Comment (FASTSHIFT) [reply demos/vector_graphic/lv_demo_vector_graphic.c:102]: Yes, the rendering behavior of both ThorVG and VG-Lite (Hardware GPU) is absolute coordinates relative to the draw buffer.

File: src/draw/vg_lite/lv_vg_lite_grad.c [hunk 2]
```diff
@@ -79,235 +74,284 @@ static vg_lite_gradient_spreadmode_t lv_spread_to_vg(lv_vector_gradient_spread_t
...
+                LV_PROFILER_END_TAG("vg_lite_draw_grad");
+            }
+            break;
+        case GRAD_TYPE_LINEAR_EXT: {
+                vg_lite_ext_linear_gradient_t * linear_grad = &grad_item->vg.linear_ext;
+                vg_lite_matrix_t * grad_mat_p = vg_lite_get_linear_grad_matrix(linear_grad);
+                LV_ASSERT_NULL(grad_mat_p);
+                *grad_mat_p = *grad_matrix;
+
+                LV_VG_LITE_ASSERT_SRC_BUFFER(&linear_grad->image);
```
Comment (liamHowatt) [inline src/draw/vg_lite/lv_vg_lite_grad.c:179]: This assertion fails for me with `LV_VG_LITE_USE_ASSERT` as `1`. I have `LV_DRAW_BUF_ALIGN` and `LV_VG_LITE_THORVG_BUF_ADDR_ALIGN` as `64`. I have `LV_VG_LITE_THORVG_16PIXELS_ALIGN` as `1`.

Edit: when running `lv_demo_vector_graphic`.

**Review by liamHowatt — COMMENTED** [review]
Initially I thought this was refactoring the new SW rendered linear, radial, conical gradients but now I understand it's solely for the vector draw interface for linear and radial gradients.

File: src/draw/vg_lite/lv_vg_lite_grad.c (hunk 2 above)
Comment (FASTSHIFT) [reply src/draw/vg_lite/lv_vg_lite_grad.c:179]: The hardware that supports `LINEAR_EXT` is a newer version of GPU. The newer version of GPU has removed the 16px alignment width requirement.

So if `LV_VG_LITE_THORVG_LINEAR_GRADIENT_EXT_SUPPORT` is enabled, please disable `LV_VG_LITE_THORVG_16PIXELS_ALIGN`.

Comment (FASTSHIFT) [reply src/draw/vg_lite/lv_vg_lite_grad.c:179]: If you set `LV_DRAW_BUF_ALIGN` to 64, you also need to align SDL's drawing buffer to 64 bytes, fix 
 here: https://github.com/lvgl/lvgl/pull/6386 .

**Comment by FASTSHIFT** [comment]
> Initially I thought this was refactoring the new SW rendered linear, radial, conical gradients but now I understand it's solely for the vector draw interface for linear and radial gradients.

@liamHowatt 
Yes, this and SW's gradient are two rendering systems and have little impact.

**Comment by FASTSHIFT** [comment]
@liamHowatt 
Is there anything else need to change? Are we ready to consider a merger? :)

**Comment by FASTSHIFT** [comment]
@kisvegabor @liamHowatt 
Is there anything else need to change? Are we ready to consider a merger? :)

**Review by kisvegabor — APPROVED** [review]

**Comment by kisvegabor** [comment]
Okay for me

File: src/draw/lv_draw_vector.c [hunk 3]
```diff
@@ -666,25 +655,40 @@ void lv_vector_dsc_set_fill_image(lv_vector_dsc_t * dsc, const lv_draw_image_dsc
...
+{
     dsc->current_dsc.fill_dsc.gradient.spread = spread;
-    lv_memcpy(&(dsc->current_dsc.fill_dsc.gradient.grad), grad, sizeof(lv_grad_dsc_t));
+}
+
+void lv_vector_dsc_set_fill_gradient_color_stops(lv_vector_dsc_t * dsc, const lv_gradient_stop_t * stops,
+                                                 uint16_t count)
+{
+    if(count > LV_GRADIENT_MAX_STOPS) {
+        LV_LOG_WARN("Too many gradient stops. Truncating to %d", LV_GRADIENT_MAX_STOPS);
```
Comment (XuNeo) [inline src/draw/lv_draw_vector.c:686]:
```suggestion
        LV_LOG_WARN("Gradient stops truncated %d", count);
```

File: src/draw/lv_draw_vector.c [hunk 4]
```diff
@@ -754,25 +758,40 @@ void lv_vector_dsc_set_stroke_miter_limit(lv_vector_dsc_t * dsc, uint16_t miter_
...
 
-void lv_vector_dsc_set_stroke_linear_gradient(lv_vector_dsc_t * dsc, const lv_grad_dsc_t * grad,
-                                              lv_vector_gradient_spread_t spread)
+void lv_vector_dsc_set_stroke_linear_gradient(lv_vector_dsc_t * dsc, float x1, float y1, float x2, float y2)
 {
     dsc->current_dsc.stroke_dsc.style = LV_VECTOR_DRAW_STYLE_GRADIENT;
     dsc->current_dsc.stroke_dsc.gradient.style = LV_VECTOR_GRADIENT_STYLE_LINEAR;
-    dsc->current_dsc.stroke_dsc.gradient.spread = spread;
-    lv_memcpy(&(dsc->current_dsc.stroke_dsc.gradient.grad), grad, sizeof(lv_grad_dsc_t));
+    dsc->current_dsc.stroke_dsc.gradient.x1 = x1;
```
Comment (XuNeo) [inline src/draw/lv_draw_vector.c:765]: Let's make the code prettier by defining a var for `stroke_dsc` and `gradient `

stroke_dsc->style = LV_VECTOR_DRAW_STYLE_GRADIENT
gradient->style = LV_VECTOR_GRADIENT_STYLE_LINEAR;
gradient->x1 = x1;
...

File: src/draw/lv_draw_vector.c [hunk 5]
```diff
@@ -754,25 +758,40 @@ void lv_vector_dsc_set_stroke_miter_limit(lv_vector_dsc_t * dsc, uint16_t miter_
...
+    dsc->current_dsc.stroke_dsc.gradient.y1 = y1;
+    dsc->current_dsc.stroke_dsc.gradient.x2 = x2;
+    dsc->current_dsc.stroke_dsc.gradient.y2 = y2;
 }
 
-void lv_vector_dsc_set_stroke_radial_gradient(lv_vector_dsc_t * dsc, const lv_grad_dsc_t * grad, float cx, float cy,
-                                              float radius, lv_vector_gradient_spread_t spread)
+void lv_vector_dsc_set_stroke_radial_gradient(lv_vector_dsc_t * dsc, float cx, float cy, float radius)
 {
     dsc->current_dsc.stroke_dsc.style = LV_VECTOR_DRAW_STYLE_GRADIENT;
```
Comment (XuNeo) [inline src/draw/lv_draw_vector.c:773]: ditto.

File: src/draw/lv_draw_vector.h [hunk 6]
```diff
@@ -391,24 +396,39 @@ void lv_vector_dsc_set_fill_image(lv_vector_dsc_t * dsc, const lv_draw_image_dsc
...
+/**
+ * Set fill gradient color stops for descriptor
  * @param dsc              pointer to a vector graphic descriptor
- * @param grad             pointer to a `lv_grad_dsc_t` variable
- * @param cx               the x for center of the circle
- * @param cy               the y for center of the circle
- * @param radius           the radius for circle
- * @param spread           the gradient spread to be set in lv_vector_gradient_spread_t format
+ * @param stops            an array of `lv_gradient_stop_t` variables
+ * @param count            the number of stops in the array
```
Comment (XuNeo) [inline src/draw/lv_draw_vector.h:428]: Specify the allowed parameter range here. 0 .. LV_GRADIENT_MAX_STOPS

**Review by XuNeo — APPROVED** [review]

File: src/draw/lv_draw_vector.c (hunk 4 above)
Comment (FASTSHIFT) [reply src/draw/lv_draw_vector.c:765]: I plan to do this during the next refactoring, but for now I will keep it consistent with the old code.

File: src/draw/lv_draw_vector.c (hunk 3 above)
Comment (FASTSHIFT) [reply src/draw/lv_draw_vector.c:686]: Fix here: https://github.com/lvgl/lvgl/pull/6399

**Pull request merged**